### PR TITLE
[WIP] 3072-bit MuHash based hash_serialized

### PR DESCRIPTION
**Pull request by sipa**
This implements a 3072-bit MuHash discussed on https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-May/014337.html as a replacement for the hash_serialized field in `gettxoutsetinfo`.

This is an order-independent hash, allowing it to be computed either by iterating over the UTXO set in non-sorted order. It also supports incremental addition and deletion of entries from the hash, allowing it to be updated on the fly for each block. Neither of these approaches is currently implemented.

### CODE REVIEW

File: src/crypto/muhash.cpp [hunk 1]
```diff
@@ -0,0 +1,329 @@
+#include "muhash.h"
```
Comment (paveljanik) [inline src/crypto/muhash.cpp]: Copyright header missing

File: src/test/crypto_tests.cpp [hunk 2]
```diff
@@ -507,4 +508,58 @@ BOOST_AUTO_TEST_CASE(countbits_tests)
...
+}
+
+BOOST_AUTO_TEST_CASE(muhash_tests)
+{
+    unsigned char out[384];
+
+    for (int i = 0; i < 10; ++i) {
+        unsigned char res[384];
+        int table[4];
+        for (int i = 0; i < 4; ++i) {
```
Comment (paveljanik) [inline src/test/crypto_tests.cpp]: `i` shadowing.

File: src/test/crypto_tests.cpp [hunk 3]
```diff
@@ -507,4 +508,58 @@ BOOST_AUTO_TEST_CASE(countbits_tests)
...
+            table[i] = insecure_rand_ctx.randbits(3);
+        }
+        for (int order = 0; order < 4; ++order) {
+            MuHash3072 acc;
+            for (int i = 0; i < 4; ++i) {
```
Comment (paveljanik) [inline src/test/crypto_tests.cpp]: `i` shadowing

File: src/test/crypto_tests.cpp [hunk 4]
```diff
@@ -507,4 +508,58 @@ BOOST_AUTO_TEST_CASE(countbits_tests)
...
+
+        MuHash3072 x = FromInt(insecure_rand_ctx.randbits(4)); // x=X
+        MuHash3072 y = FromInt(insecure_rand_ctx.randbits(4)); // x=X
+        MuHash3072 z; // x=X, y=Y, z=1
+        z *= x; // x=X, y=Y, z=X
+        z /= y; // x=X, y=Y, z=X/Y
+        y /= x; // x=X, y=Y/X, z=X/Y
+        z *= y; // x=X, y=Y/X, z=1
+        z.Finalize(out);
+        for (int i = 0; i < 384; ++i) {
```
Comment (paveljanik) [inline src/test/crypto_tests.cpp]: `i` shadowing

**Review by paveljanik — COMMENTED** [review]
Release notes entry needed.

File: src/crypto/muhash.cpp (hunk 1 above)
Comment (sipa) [inline src/crypto/muhash.cpp]: Fixed.

File: src/test/crypto_tests.cpp (hunk 2 above)
Comment (sipa) [inline src/test/crypto_tests.cpp]: Fixed.

File: src/test/crypto_tests.cpp (hunk 4 above)
Comment (sipa) [inline src/test/crypto_tests.cpp]: Fixed.

**Comment by n1bor** [comment]
@sipa - did this POC of chainstate only sync:
https://gist.github.com/n1bor/d5b0330a9addb0bf5e0f869518883522
https://botbot.me/freenode/bitcoin-core-dev/2016-10-24/?msg=75398575&page=3

Worked well - syncs as fast at you can download chainstate db (30mins for 2.8gig chainstate). Network was the bottleneck.
Should be able to rework to use MuHash easily - and rethink the snapshot interval/process.
But if consensus is to NEVER commit chain-state to blockchain - than is all a little pointless. gmaxwell did suggest an expiring consensus rule so if want to change from MuHash would be easy - which would do.

**Comment by sipa** [comment]
Closing as stale.

**Pull request closed without merging**